In [1]:
%cd ..

D:\SoftUni\Data Science\Project2


### Imports

In [2]:
import pandas as pd

import numpy as np

from src import functions

In this notebook all preprocessed datasets are merged into final working dataset.

### Reading datasets

In [3]:
drugs_distribution_data = pd.read_csv('data/preprocessed_datasets/drugs_distribution_by_country.csv')

In [4]:
drugs_distribution_data

country  heroin_trade  cocaine_trade  \
0                            Turkey           8.0            4.0   
1                        Cabo Verde           4.0            7.0   
2                      South Africa           7.5            5.5   
3                          Colombia           5.0            9.5   
4                              Peru           3.5            9.0   
..                              ...           ...            ...   
188                         Jamaica           3.0            6.0   
189  St. Vincent and the Grenadines           1.0            7.5   
190                        Slovenia           4.0            4.5   
191                       St. Lucia           1.0            7.0   
192                        Dominica           1.0            6.0   

     cannabis_trade  synthetic_drug_trade  
0               5.0                   5.5  
1               5.0                   4.5  
2               3.5                   8.5  
3               8.0                   5.0  
4               5.5                   4.0  
..              ...                   ...  
188             9.0                   3.0  
189             7.5                   1.0  
190             5.0                   5.5  
191             7.0                   1.0  
192             5.0                   2.5  

[193 rows x 5 columns]

In [5]:
mental_disorders_data = pd.read_csv('data/preprocessed_datasets/mental_disorders_by_country.csv')

In [6]:
mental_disorders_data

country  anxiety_disorders_rate  adh_disorders_rate  \
0             Afghanistan                  6036.0              1006.0   
1                  Angola                  4124.0               484.0   
2                 Albania                  4621.0              1014.0   
3                 Andorra                  6651.0              1291.0   
4    United Arab Emirates                  5168.0               448.0   
..                    ...                     ...                 ...   
199                Zambia                  4505.0               487.0   
200              Zimbabwe                  3913.0               485.0   
201                 Egypt                  5214.0              1010.0   
202                 Sudan                  5734.0              1002.0   
203                 China                  3482.0              2184.0   

     biploar_disorders_rate  depressive_disorders_rate  schizophrenia_rate  
0                     711.0                     5916.0               218.0  
1                     502.0                     6772.0               224.0  
2                     480.0                     2977.0               221.0  
3                     820.0                     4495.0               271.0  
4                     713.0                     4332.0               266.0  
..                      ...                        ...                 ...  
199                   549.0                     4911.0               217.0  
200                   500.0                     3910.0               208.0  
201                   687.0                     4605.0               243.0  
202                   715.0                     5035.0               231.0  
203                   173.0                     2876.0               312.0  

[204 rows x 6 columns]

In [7]:
suds_data = pd.read_csv('data/preprocessed_datasets/sud_disorders_by_country.csv')

In [8]:
suds_data

country  amphetamine_disorders_rate  \
0             Afghanistan                        32.0   
1                  Angola                        38.0   
2                 Albania                       131.0   
3                 Andorra                       130.0   
4    United Arab Emirates                        53.0   
..                    ...                         ...   
199                Zambia                        37.0   
200              Zimbabwe                        28.0   
201                 Egypt                        34.0   
202                 Sudan                        32.0   
203                 China                       267.0   

     cannabis_disorders_rate  cocaine_disorders_rate  opioid_disorders_rate  \
0                      130.0                    14.0                  249.0   
1                      177.0                     8.0                   72.0   
2                      246.0                    28.0                  134.0   
3                      468.0                   129.0                  112.0   
4                      138.0                    18.0                  371.0   
..                       ...                     ...                    ...   
199                    174.0                     6.0                   68.0   
200                    208.0                    27.0                  145.0   
201                    112.0                    12.0                  199.0   
202                    129.0                    13.0                  178.0   
203                    207.0                     6.0                   94.0   

     overal_suds_rate  
0               413.0  
1               277.0  
2               553.0  
3               789.0  
4               732.0  
..                ...  
199             290.0  
200             408.0  
201             368.0  
202             378.0  
203             540.0  

[204 rows x 6 columns]

### Merging datasets containing data for diffrent mental disorders rates and substance use disorders rates.

In [9]:
disorders_data = mental_disorders_data.merge(suds_data, on = 'country', how = 'inner')

In [10]:
disorders_data

country  anxiety_disorders_rate  adh_disorders_rate  \
0             Afghanistan                  6036.0              1006.0   
1                  Angola                  4124.0               484.0   
2                 Albania                  4621.0              1014.0   
3                 Andorra                  6651.0              1291.0   
4    United Arab Emirates                  5168.0               448.0   
..                    ...                     ...                 ...   
199                Zambia                  4505.0               487.0   
200              Zimbabwe                  3913.0               485.0   
201                 Egypt                  5214.0              1010.0   
202                 Sudan                  5734.0              1002.0   
203                 China                  3482.0              2184.0   

     biploar_disorders_rate  depressive_disorders_rate  schizophrenia_rate  \
0                     711.0                     5916.0               218.0   
1                     502.0                     6772.0               224.0   
2                     480.0                     2977.0               221.0   
3                     820.0                     4495.0               271.0   
4                     713.0                     4332.0               266.0   
..                      ...                        ...                 ...   
199                   549.0                     4911.0               217.0   
200                   500.0                     3910.0               208.0   
201                   687.0                     4605.0               243.0   
202                   715.0                     5035.0               231.0   
203                   173.0                     2876.0               312.0   

     amphetamine_disorders_rate  cannabis_disorders_rate  \
0                          32.0                    130.0   
1                          38.0                    177.0   
2                         131.0                    246.0   
3                         130.0                    468.0   
4                          53.0                    138.0   
..                          ...                      ...   
199                        37.0                    174.0   
200                        28.0                    208.0   
201                        34.0                    112.0   
202                        32.0                    129.0   
203                       267.0                    207.0   

     cocaine_disorders_rate  opioid_disorders_rate  overal_suds_rate  
0                      14.0                  249.0             413.0  
1                       8.0                   72.0             277.0  
2                      28.0                  134.0             553.0  
3                     129.0                  112.0             789.0  
4                      18.0                  371.0             732.0  
..                      ...                    ...               ...  
199                     6.0                   68.0             290.0  
200                    27.0                  145.0             408.0  
201                    12.0                  199.0             368.0  
202                    13.0                  178.0             378.0  
203                     6.0                   94.0             540.0  

[204 rows x 11 columns]

### Preprocessing 'country' column

The goal is to merge 'disorders' and 'drugs distribution' datasets by their common column - 'country'.

In [11]:
countries_in_drugs_data = np.sort(drugs_distribution_data['country'].unique())

In [12]:
len(countries_in_drugs_data)

193

In [13]:
countrues_in_disorders_data = np.sort(disorders_data['country'].unique())

In [14]:
len(countrues_in_disorders_data)

204

Countries in 'disorders_data' are 11 more than in 'drugs_distribution_data'. 
My approach here will be to drop the countries in each dataset which are not present in the another. At the other hand there are some differences between the spelling of the same countries in both datasets, which requires additional prepocessing.

In [15]:
# Country names which are present in drugs distribution dataset but not in the disorders dataset
different_names_in_drugs = np.setdiff1d(countries_in_drugs_data, countrues_in_disorders_data)

In [16]:
different_names_in_drugs

array(['Bolivia', 'Brunei', 'Congo, Dem. Rep.', 'Congo, Rep.',
       'Czech Republic', 'Iran', 'Korea, DPR', 'Korea, Rep.', 'Laos',
       'Liechtenstein', 'Moldova', 'Russia', 'St. Kitts and Nevis',
       'St. Lucia', 'St. Vincent and the Grenadines', 'Syria', 'Tanzania',
       'Turkey', 'United States', 'Venezuela', 'Vietnam'], dtype=object)

In [17]:
# Country names which are present in the disorders dataset but not in the drugs distribution dataset
np.setdiff1d(countrues_in_disorders_data, countries_in_drugs_data)

array(['American Samoa', 'Bermuda', 'Bolivia (Plurinational State of)',
       'Brunei Darussalam', 'Congo', 'Cook Islands', 'Czechia',
       "Democratic People's Republic of Korea",
       'Democratic Republic of the Congo', 'Greenland', 'Guam',
       'Iran (Islamic Republic of)', "Lao People's Democratic Republic",
       'Niue', 'Northern Mariana Islands', 'Palestine', 'Puerto Rico',
       'Republic of Korea', 'Republic of Moldova', 'Russian Federation',
       'Saint Kitts and Nevis', 'Saint Lucia',
       'Saint Vincent and the Grenadines', 'Syrian Arab Republic',
       'Taiwan (Province of China)', 'Tokelau', 'Türkiye',
       'United Republic of Tanzania', 'United States Virgin Islands',
       'United States of America', 'Venezuela (Bolivarian Republic of)',
       'Viet Nam'], dtype=object)

We can observe that data for Liechtenstein as a country is present only in drugs dataframe. Also there are several countries that are present only in mental disorders dataset. The merge process will 'clean' all of them from the resulting dataframe. The only remain for preprocessing countries will be these with different spelled names. A little work of comparing the above arrays is required in order to create an array which contains all different spelled country names in disordes dataset. Then it will be easy to equalize them to the corresponding countries in drugs dataframe and to merge them after.

In [18]:
different_spelled_countries_in_disorders = np.array(['Bolivia (Plurinational State of)', 'Brunei Darussalam', 'Democratic Republic of the Congo', 'Congo', 'Czechia',
                    'Iran (Islamic Republic of)', 'Democratic People\'s Republic of Korea', 'Republic of Korea', 'Lao People\'s Democratic Republic',
                    'Republic of Moldova', 'Russian Federation','Saint Kitts and Nevis', 'Saint Lucia', 'Saint Vincent and the Grenadines',
                    'Syrian Arab Republic', 'United Republic of Tanzania', 'Türkiye','United States of America', 'Venezuela (Bolivarian Republic of)', 'Viet Nam'])

In [19]:
# Delete Liechtenstein in the array
index = np.argwhere(different_names_in_drugs == 'Liechtenstein')
different_names_in_drugs = np.delete(different_names_in_drugs, index)

Now I can create a dictionary with key - value pairs corresponding to all different spelled countries in both dataframes.

In [20]:
different_named_countries_dict = {k: v for k, v in zip(different_spelled_countries_in_disorders, different_names_in_drugs)}

In [21]:
different_named_countries_dict

{'Bolivia (Plurinational State of)': 'Bolivia',
 'Brunei Darussalam': 'Brunei',
 'Democratic Republic of the Congo': 'Congo, Dem. Rep.',
 'Congo': 'Congo, Rep.',
 'Czechia': 'Czech Republic',
 'Iran (Islamic Republic of)': 'Iran',
 "Democratic People's Republic of Korea": 'Korea, DPR',
 'Republic of Korea': 'Korea, Rep.',
 "Lao People's Democratic Republic": 'Laos',
 'Republic of Moldova': 'Moldova',
 'Russian Federation': 'Russia',
 'Saint Kitts and Nevis': 'St. Kitts and Nevis',
 'Saint Lucia': 'St. Lucia',
 'Saint Vincent and the Grenadines': 'St. Vincent and the Grenadines',
 'Syrian Arab Republic': 'Syria',
 'United Republic of Tanzania': 'Tanzania',
 'Türkiye': 'Turkey',
 'United States of America': 'United States',
 'Venezuela (Bolivarian Republic of)': 'Venezuela',
 'Viet Nam': 'Vietnam'}

The dictionary is used to change the misspelled country names in disorders dataset.

In [22]:
functions.rplace_col_values(dict = different_named_countries_dict, df = disorders_data, col = 'country')

D:\SoftUni\Data Science\Project2\src\functions.py:22: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].replace(df[df[col] == key][col].values[0], value, inplace = True)


In [23]:
disorders_data['country'].unique()

array(['Afghanistan', 'Angola', 'Albania', 'Andorra',
       'United Arab Emirates', 'Argentina', 'Armenia', 'American Samoa',
       'Antigua and Barbuda', 'Australia', 'Austria', 'Azerbaijan',
       'Burundi', 'Belgium', 'Benin', 'Burkina Faso', 'Bangladesh',
       'Bulgaria', 'Bahrain', 'Bahamas', 'Bosnia and Herzegovina',
       'Belarus', 'Belize', 'Bermuda', 'Bolivia', 'Brazil', 'Barbados',
       'Brunei', 'Bhutan', 'Botswana', 'Central African Republic',
       'Canada', 'Switzerland', 'Chile', "Côte d'Ivoire", 'Cameroon',
       'Congo, Dem. Rep.', 'Congo, Rep.', 'Cook Islands', 'Colombia',
       'Comoros', 'Cabo Verde', 'Costa Rica', 'Cuba', 'Cyprus',
       'Czech Republic', 'Germany', 'Djibouti', 'Dominica', 'Denmark',
       'Dominican Republic', 'Algeria', 'Ecuador', 'Eritrea', 'Spain',
       'Estonia', 'Ethiopia', 'Finland', 'Fiji', 'France',
       'Micronesia (Federated States of)', 'Gabon', 'United Kingdom',
       'Georgia', 'Ghana', 'Guinea', 'Gambia', 'Guinea-B

### Merging disorders and drugs distribution datasets

Now I can merge the data from both dataframes to get a resulted dataframe with 192 observations(number of countries in drugs data - 1 for Liechtenstein) and 15 features.

In [24]:
drugs_disorders_data = drugs_distribution_data.merge(disorders_data, on = "country", how = 'inner')

In [25]:
drugs_disorders_data.head(2)

country  heroin_trade  cocaine_trade  cannabis_trade  \
0      Turkey           8.0            4.0             5.0   
1  Cabo Verde           4.0            7.0             5.0   

   synthetic_drug_trade  anxiety_disorders_rate  adh_disorders_rate  \
0                   5.5                  5420.0              1375.0   
1                   4.5                  3496.0               505.0   

   biploar_disorders_rate  depressive_disorders_rate  schizophrenia_rate  \
0                   737.0                     4730.0               246.0   
1                   473.0                     5801.0               241.0   

   amphetamine_disorders_rate  cannabis_disorders_rate  \
0                        18.0                     82.0   
1                        35.0                    128.0   

   cocaine_disorders_rate  opioid_disorders_rate  overal_suds_rate  
0                    15.0                  182.0             323.0  
1                     8.0                   73.0             291.0

In [26]:
drugs_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 192 entries, 0 to 191
Data columns (total 15 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   country                     192 non-null    object 
 1   heroin_trade                192 non-null    float64
 2   cocaine_trade               192 non-null    float64
 3   cannabis_trade              192 non-null    float64
 4   synthetic_drug_trade        192 non-null    float64
 5   anxiety_disorders_rate      192 non-null    float64
 6   adh_disorders_rate          192 non-null    float64
 7   biploar_disorders_rate      192 non-null    float64
 8   depressive_disorders_rate   192 non-null    float64
 9   schizophrenia_rate          192 non-null    float64
 10  amphetamine_disorders_rate  192 non-null    float64
 11  cannabis_disorders_rate     192 non-null    float64
 12  cocaine_disorders_rate      192 non-null    float64
 13  opioid_disorders_rate       192 non

In [27]:
# Changing the order of columns to more symetric way - drugs distribution rates in the middle with mental and substance use disorders on both sides
drugs_disorders_data = drugs_disorders_data[[
    'country',  
    'opioid_disorders_rate', 
    'cocaine_disorders_rate', 
    'cannabis_disorders_rate', 
    'amphetamine_disorders_rate', 
    'heroin_trade', 
    'cocaine_trade', 
    'cannabis_trade', 
    'synthetic_drug_trade', 
    'overal_suds_rate', 
    'anxiety_disorders_rate', 
    'adh_disorders_rate', 
    'biploar_disorders_rate', 
    'depressive_disorders_rate', 
    'schizophrenia_rate',
]]

In [28]:
drugs_disorders_data

country  opioid_disorders_rate  \
0                            Turkey                  182.0   
1                        Cabo Verde                   73.0   
2                      South Africa                  133.0   
3                          Colombia                  100.0   
4                              Peru                   99.0   
..                              ...                    ...   
187                         Jamaica                   92.0   
188  St. Vincent and the Grenadines                   91.0   
189                        Slovenia                  214.0   
190                       St. Lucia                   95.0   
191                        Dominica                   86.0   

     cocaine_disorders_rate  cannabis_disorders_rate  \
0                      15.0                     82.0   
1                       8.0                    128.0   
2                     111.0                    324.0   
3                     121.0                    576.0   
4                      52.0                    206.0   
..                      ...                      ...   
187                   144.0                    625.0   
188                   115.0                    586.0   
189                    50.0                    462.0   
190                    90.0                    677.0   
191                   180.0                    704.0   

     amphetamine_disorders_rate  heroin_trade  cocaine_trade  cannabis_trade  \
0                          18.0           8.0            4.0             5.0   
1                          35.0           4.0            7.0             5.0   
2                         141.0           7.5            5.5             3.5   
3                          41.0           5.0            9.5             8.0   
4                          70.0           3.5            9.0             5.5   
..                          ...           ...            ...             ...   
187                        48.0           3.0            6.0             9.0   
188                        50.0           1.0            7.5             7.5   
189                       109.0           4.0            4.5             5.0   
190                        51.0           1.0            7.0             7.0   
191                        53.0           1.0            6.0             5.0   

     synthetic_drug_trade  overal_suds_rate  anxiety_disorders_rate  \
0                     5.5             323.0                  5420.0   
1                     4.5             291.0                  3496.0   
2                     8.5             775.0                  4826.0   
3                     5.0             911.0                  5601.0   
4                     4.0             474.0                  7241.0   
..                    ...               ...                     ...   
187                   3.0            1013.0                  5211.0   
188                   1.0             848.0                  5195.0   
189                   5.5             655.0                  4076.0   
190                   1.0             936.0                  5465.0   
191                   2.5            1072.0                  5226.0   

     adh_disorders_rate  biploar_disorders_rate  depressive_disorders_rate  \
0                1375.0                   737.0                     4730.0   
1                 505.0                   473.0                     5801.0   
2                 498.0                   507.0                     5310.0   
3                1549.0                   718.0                     2460.0   
4                2183.0                   835.0                     2697.0   
..                  ...                     ...                        ...   
187              3073.0                   844.0                     3491.0   
188              3094.0                   830.0                     3849.0   
189              1011.0                   481.0                     3693.0   
190             

### Saving final dataset

In [29]:
drugs_disorders_data.to_csv('data/final_merged_dataset/drugs_disorders_by_country.csv', index = False)